File: C/data_trainer.c
// Libraries
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

// global variables
#define NULL_CHAR '\0'

// structs
typedef struct TreeNode {
    char letter;
    int frequencyCount;
    struct TreeNode* children;
    struct TreeNode* next;
} TreeNode;

typedef struct FrequencyTree {
    TreeNode* root;
} FrequencyTree;

typedef struct FlatNode {
    char letter;
    int depth;
    int count;
    char parent;   
} FlatNode;

// function definitions
TreeNode* createTreeNode(char letter);
FrequencyTree* createFrequencyTree();
void insertWord(FrequencyTree* tree, char* word);
void printTree(TreeNode* node, char* prefix, int is_last_sibling);
void writeTreeToFile(TreeNode* node, FILE* file, char* prefix, int is_last_sibling);
void freeTreeNode(TreeNode* node);
void freeFrequencyTree(FrequencyTree* tree);
FlatNode* flattenTree(FrequencyTree* tree, char traversal_method);
int countNodes(TreeNode* node);
void dfs(TreeNode* node, int depth, char parent, FlatNode* output_list, int* index);
void bfs(TreeNode* rootNode, FlatNode* output_list);


// main
int main() {
    // open file with 370k words
    FILE* file = fopen("../Training Files/words_alpha.txt", "r");
    if (file == NULL) {
        printf("Error opening file.\n");
        return 1;
    }

    // create frequency tree
    FrequencyTree* tree = createFrequencyTree();

    // read words from file
    char word[100];
    while (fscanf(file, "%s", word) != EOF) {
        insertWord(tree, word);
    }
    fclose(file);

    // flatten the tree
    FlatNode* flatList = flattenTree(tree, 'd');

    // print flat list
    int totalNodes = countNodes(tree->root);
    for (int index = 0; index < totalNodes; index++)
    {
        printf("Letter: %c, Depth: %d, Count: %d, Parent: %c\n",
            flatList[index].letter, flatList[index].depth, flatList[index].count, flatList[index].parent);
    } 

    // Print entire tree in terminal
    // printf("Frequency Tree:\n");
    // printTree(tree->root, "", 1);

    // Write entire tree to file
    // file = fopen("frequency_tree_output_C_Version.txt", "w");
    // if (file == NULL) {
    //     printf("Error opening file.\n");
    //     return 1;
    // }
    // writeTreeToFile(tree->root, file, "", 1);
    // fclose(file);

    // Free the entire tree and then the tree itself
    free(flatList);
    freeFrequencyTree(tree);

    return 0;
}

// function implementations
TreeNode* createTreeNode(char letter)
{
    TreeNode* newNode = (TreeNode*)malloc(sizeof(TreeNode));
    newNode->letter = letter;
    newNode->frequencyCount = 1;
    newNode->children = NULL;
    newNode->next = NULL;
    return newNode;
}

FrequencyTree* createFrequencyTree()
{
    FrequencyTree* tree = (FrequencyTree*) malloc(sizeof(FrequencyTree));
    tree->root = createTreeNode(NULL_CHAR);
    return tree;
}

void insertWord(FrequencyTree* tree, char* word)
{
    int letter;
    char charToInsert;
    
    TreeNode* currentNode = tree->root;
    
    for (letter = 0; word[letter] != NULL_CHAR; letter++) 
    {
        charToInsert = word[letter];
        
        TreeNode* childNode = currentNode->children;

        while (childNode != NULL) 
        {
            if (childNode->letter == charToInsert) 
            {
                childNode->frequencyCount++;
                break;
            }
            childNode = childNode->next;
        }
        
        if (childNode == NULL) {
            TreeNode* newChild = createTreeNode(charToInsert);
            newChild->next = currentNode->children;
            currentNode->children = newChild;
        }
        currentNode = currentNode->children;
    }
}

void printTree(TreeNode* node, char* prefix, int is_last_sibling) 
{
    if (node == NULL) return;

    char branch_symbol[5];
    strcpy(branch_symbol, is_last_sibling ? "L--- " : "T--- ");
    printf("%s%s%c (count: %d)\n", prefix, branch_symbol, node->letter, node->frequencyCount);

    TreeNode* child = node->children;
    while (child != NULL) 
    {
        int new_prefix_len = strlen(prefix) + 5;
        char* new_prefix = (char*)malloc(new_prefix_len * sizeof(char));
        strcpy(new_prefix, prefix);
        strcat(new_prefix, is_last_sibling ? "    " : "|   ");
        printTree(child, new_prefix, child->next == NULL);
        free(new_prefix);
        child = child->next;
    }
}

void writeTreeToFile(TreeNode* node, FILE* file, char* prefix, int is_last_sibling) 
{
    // return if empty tree
    if (node == NULL) return;

    // Write the node to the file regardless of the prefix
    fprintf(file, "%s%s%c (count: %d)\n", prefix, is_last_sibling ? "L--- " : "T--- ", node->letter, node->frequencyCount);

    // allocate memory for children prefixes
    char* childPrefix = (char*)malloc(strlen(prefix) + 5);
    strcpy(childPrefix, prefix);
    strcat(childPrefix, is_last_sibling ? "    " : "|   "); // set spacing

    // loop through children nodes
    TreeNode* child = node->children;
    while (child != NULL) 
    {
        // write child node
        writeTreeToFile(child, file, childPrefix, child->next == NULL);
        child = child->next;
    }
    free(childPrefix); // free child prefix
}

void freeTreeNode(TreeNode* node) {
    if (node == NULL) return;

    // Recursively free all children of the current node
    TreeNode* child = node->children;
    while (child != NULL) {
        TreeNode* nextChild = child->next; // Save the next child before freeing
        freeTreeNode(child);
        child = nextChild;
    }

    // Free the current node
    free(node);
}

void freeFrequencyTree(FrequencyTree* tree) {
    if (tree == NULL) return;

    // Free the entire tree
    freeTreeNode(tree->root);

    // Free the tree itself
    free(tree);
}

int countNodes(TreeNode* node) 
{
    int totalNodeCount = 1;
    TreeNode* child = node->children;
    if (node == NULL) return 0;

    while (child != NULL) 
    {
        totalNodeCount += countNodes(child);
        child = child->next;
    }

    return totalNodeCount;
}

void dfs(TreeNode* node, int depth, char parent, FlatNode* output_list, int* index)
{
    if (node == NULL) return;

    output_list[*index].letter = node->letter;
    output_list[*index].depth = depth;
    output_list[*index].count = node->frequencyCount;
    output_list[*index].parent = parent;
    (*index)++;

    TreeNode* child = node->children;
    while (child != NULL) {
        dfs(child, depth + 1, node->letter, output_list, index);
        child = child->next;
    }
}

void bfs(TreeNode* rootNode, FlatNode* output_list)
{

}

FlatNode* flattenTree(FrequencyTree* tree, char traversal_method)
{
    int index = 0, depth = 0;
    TreeNode* rootNode = tree->root;
    int totalNodes = countNodes(tree->root); // get total nodes in tree
    
    // malloc flat list
    FlatNode* flatList = (FlatNode*)malloc(totalNodes * sizeof(FlatNode));

    if (flatList == NULL)
    {
        printf("Memory allocation failed.");
        exit(EXIT_FAILURE);
    }

    // flatten tree with dfs
    if (traversal_method == 'd')
    {
        dfs(rootNode, depth, NULL_CHAR, flatList, &index);
    }

    // flatten tree with bfs
    else if (traversal_method == 'b')
    {
        bfs(rootNode, flatList);
    }

    // illegal traversal method
    else
    {
        fprintf(stderr, "Invalid traversal method. Use 'dfs' or 'bfs'.\n");
        exit(EXIT_FAILURE);
    }

    return flatList;
}

// void flattenTree(TreeNode* node, FlatNode* output_list)
// {
    // set flag variable subtreeEND = FALSE
    // set variable for iteration counter = 0

    // enter while loop to iterate over root node of data structure
    // loop over its children until rootNode->currentChild == NULL && subtreeEND == true

        // enter while loop to iterate to nodes at depth == iteration
        // (iteration 0 gets all nodes of depth 0, iteration 1 gets all nodes of depth 1, etc)

        // enter while loop to iterate over the children node at this depth
        // read contents of node, write them into a struct

            // What to write: label, depth, count, parent (NULLCHAR if depth = 0)
        
        // add struct into 1d array at index
    
        // increment index

// }

File: Python/data_trainer.py
from collections import deque
# classes to handle creation of frequency tree
class TreeNode:
    def __init__(self) -> None:
        self.letter = ""
        self.frequencyCount = 0
        self.children = dict()

class FrequencyTree:
    def __init__(self) -> None:
        self.root = TreeNode() # set root of tree
    
    def insertWord(self, word):
        currentNode = self.root # wkg pointer for root

        # loop through letters in word
        for char in word:

            # check if letter is not a child of current node
            if char not in currentNode.children:
                currentNode.children[char] = TreeNode() # add child node
                currentNode.children[char].letter = char # store letter in node
            
            # move to next letter (node) and increase frequency count of letter at node
            currentNode = currentNode.children[char]
            currentNode.frequencyCount += 1
        

    def printTree(self, node=None, prefix="", is_last_sibling=True):
        if node is None:
            node = self.root

        else:  # Printing nodes other than the root
            branch_symbol = "└── " if is_last_sibling else "├── "
            print(f"{prefix}{branch_symbol}{node.letter} (count: {node.frequencyCount})")
            prefix += "    " if is_last_sibling else "│   " # adjust output based on next nodes siblings
        
        children = list(node.children.values())
        for currentIteration, child in enumerate(children):
            # Pass the updated prefix and is_last flag based on whether the child is the last in its siblings
            self.printTree(child, prefix, currentIteration == len(children) - 1)

def writeTraversalToFile(result_list):
    #write entire result_list to file
    with open("/../Output/5_letter_frequency_list_padded.txt", "w") as output_file:
      for item in result_list:
            combined_values = []
            for i in range(0, len(item), 2):
                combined_values.append(str(item[i]) + str(item[i + 1]))
            output_file.write(",".join(combined_values) + "\n")

def traverse_tree(node, result_list, length, current_list = None):

  # set local copy of current_list
  if current_list is None:
      current_list = []
  
  # Append current node's letter and freequency to the current list
  current_list.append(node.letter)
  current_list.append(node.frequencyCount)

  # Check if the node is an end node
  if len(node.children) == 0:
    word_list = []
    word_list = current_list[2:]
    
    while length < 5:
      word_list.append('_')
      word_list.append(1)
      length+=1

    # Append the list to the result list
    result_list.append(word_list)

    return

  length+=1
  # Iterate through the children of the node
  for child_node in node.children.values():
      
      # Recursively traverse the child node
      traverse_tree(child_node, result_list, length, current_list)
      
      # Remove the current node's letter and frequency from the current list
      for i in range(2):
        current_list.pop()

def writeTraversalToFileCommaSep(result_list):
    #write entire result_list to file
    with open("/../Output/5_letter_frequency_list_padded_comma_sep.txt", "w") as output_file:
      for item in result_list:
        output_file.write(','.join(map(str, item))+"\n")

class FlatNode():
    def __init__(self, label, depth, count, parent):
        self.label = label
        self.depth = depth
        self.count = count
        self.parent = parent

def flattenTree(root_node, output_list, traversal_method='dfs'):
    if traversal_method == 'dfs':
        def dfs(node, depth, parent):
            if not node.children:
                return

            for child in node.children.values():
                output_list.append(FlatNode(child.letter, depth, child.frequencyCount, parent))
                dfs(child, depth + 1, child.letter)
        dfs(root_node, 0, None)

    elif traversal_method == 'bfs':
        queue = deque([(root_node, 0, None)])
        while queue:
            node, depth, parent = queue.popleft()
            output_list.append(FlatNode(node.letter, depth, node.frequencyCount, parent))
            for child in node.children.values():
                queue.append((child, depth + 1, node.letter))
                
    else:
        raise ValueError("Invalid traversal method. Use 'dfs' or 'bfs'.")

    return output_list


def writeTreeToFile(tree, filename):
    # Open the file in write mode with UTF-8 encoding
    with open(filename, 'w', encoding='utf-8') as file:
        def printTreeToFile(node=None, prefix="", is_last_sibling=True):
            if node is None:
                node = tree.root
            else: # Printing nodes other than the root
                branch_symbol = "└── " if is_last_sibling else "├── "
                file.write(f"{prefix}{branch_symbol}{node.letter} (count: {node.frequencyCount})\n")
                prefix += "    " if is_last_sibling else "│   " # adjust output based on next nodes siblings
            
            children = list(node.children.values())
            for currentIteration, child in enumerate(children):
                # Pass the updated prefix and is_last flag based on whether the child is the last in its siblings
                printTreeToFile(child, prefix, currentIteration == len(children) - 1)
        
        # Start the recursive function call
        printTreeToFile()

def main():
    # read all words in txt file
    with open('../Output/5_letter_dict.txt', 'r') as file:
        words = file.readlines()

    tree = FrequencyTree()

    for word in words:
        word = word.strip()
        tree.insertWord(word)

    # output_list_dfs = flattenTree(tree.root, [], 'dfs')
    # output_list_bfs = flattenTree(tree.root, [], 'bfs')

    # print flattened tree using dfs
    # print("DFS Output list")
    # for flat_node in output_list_dfs:
    #     print(f"Label: {flat_node.label}, Depth: {flat_node.depth}, Count: {flat_node.count}, Parent: {flat_node.parent}")

    # print flattened tree using bfs
    # print("BFS Output list")
    # for flat_node in output_list_bfs:
        # print(f"Label: {flat_node.label}, Depth: {flat_node.depth}, Count: {flat_node.count}, Parent: {flat_node.parent}")

    # print entire tree in terminal    
    # tree.printTree()

    # write entire tree to file
    # writeTreeToFile(tree, 'frequency_tree_output_Python_Version.txt')

    # Create an empty list to store the result
    result_list = []
    length = 0

    # Traverse the frequency tree
    traverse_tree(tree.root, [], length, result_list)

    # Print the result list
    print(result_list)

    # create 5 letter dictionary
    # with open("../Training Files/words_alpha.txt", "r") as input_file:
    #     lines = input_file.readlines()

    # with open("../Output/5_letter_dict.txt", "w") as file_out:
    #     for line in lines:
    #         words = line.split()

    #         for word in words:
    #             if len(word) <= 5:
    #                 file_out.write(word + "\n")
    #write entire result_list to file
    #writeTraversalToFile(result_list)

    writeTraversalToFileCommaSep(result_list)

if __name__ == "__main__":
    main()





